x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/CostEntry.java

```diff
@@ -14,20 +14,32 @@
 import java.util.Map;
 
 final class CostEntry implements ToXContentObject {
-    final double costFactor;
+    final double usd_per_hour;
 
-    CostEntry(double costFactor) {
-        this.costFactor = costFactor;
+    CostEntry(double usdPerHour) {
+        this.usd_per_hour = usdPerHour;
     }
 
     public static CostEntry fromSource(Map<String, Object> source) {
-        return new CostEntry((Double) source.get("cost_factor"));
+        var val = source.get("usd_per_hour");
+
+        if (val instanceof Double d) {
+            return new CostEntry(d);
+        }
+
+        // Some JSON values have no decimal places and are passed in as Integers.
```

Comment:
Can we address this inconsistency instead when we write the data? This would shift complexity to the code that executes once (writing) from the code that executes frequently (reading).

If we can rely on the fact that `cost_factor` is either a double value or `null` we can simplify this method to:

```java
return new CostEntry((Double) source.getOrDefault("cost_factor", CostCalculator.DEFAULT_COST_USD_PER_CORE_HOUR * HostMetadata.DEFAULT_PROFILING_NUM_CORES));
```

(you could extract the multiplication to a class constant and simplify this further)

Reply:
> Can we address this inconsistency instead when we write the data?

That was my first consideration and I decided to not rely on the writer.
The current generation of the data is temporary and likely (hopefully) will change in the future. But whatever we decide for, the danger of not generating superfluous "." or ".0" is still there - every programming language or string format library may handle this differently. Especially, when we open the data for manual changes, it is much safer to handle this on the reader side. Not to forget, it's the Java code that is so picky here - better not export this behavior to the data generators.

Reply:
I gave it some thought and IMHO we should at least check that the value is either numeric or `null` and reject all other values. How about something like this?

```java
    public static CostEntry fromSource(Map<String, Object> source) {
        var val = source.get("usd_per_hour");

        if (val instanceof Number n) {
            // Some JSON values have no decimal places and are passed in as Integers.
            return new CostEntry(n.doubleValue());
        } else if (val == null) {
            return new CostEntry(CostCalculator.DEFAULT_COST_USD_PER_CORE_HOUR * HostMetadata.DEFAULT_PROFILING_NUM_CORES);
        } else {
            throw new IllegalArgumentException("[" + val + "] is an invalid value for [usd_per_hour]");
        }
    }
```